### ISSUE TITLE
How to include essential boundary conditions for saddle point problem?

### ISSUE THREAD

**Issue by AIBSCT80** [mfem/mfem#4418]
Hi, 

I have an issues assigning essential boundary conditions to the saddle point problem

```
A       B
B^T  C
```

In order to include essential boundary conditions  for the primal and dual,  I called the following:

```
      mVarf->FormSystemMatrix(u_ess_tdof_list, A);
      bVarf->FormRectangularSystemMatrix(u_ess_tdof_list, p_ess_tdof_list, B);
      pVarf->FormSystemMatrix(p_ess_tdof_list, C);
```

However, my p_ess_tdof_list is empty since I will not need to assign essential boundary 
conditions to C. But in order to incorporating boundary conditions into operator B, 
I need to call 

` bVarf->FormRectangularSystemMatrix(u_ess_tdof_list, p_ess_tdof_list, B);`

even the p_ess_tdof_list is empty.  However, I always get zero boundary conditions for 
p variables.  This is confusing for me.  Many thanks.

**Comment by psocratis** [comment]
@AIBSCT80 you can take a look [here](https://github.com/mfem/mfem/blob/ada5f0cada1d95e0f404b3ea6e0e2e1157eaf16c/fem/complex_fem.cpp#L400) how the complex-arithmetic systems are assembled and BCs are enforced on both real and imaginary part. In this case the assembled system looks like this 

$$\begin{bmatrix}
A_r & -A_i \\
A_i & A_r
\end{bmatrix}\begin{bmatrix}
x_r  \\
x_i 
\end{bmatrix}=\begin{bmatrix}
y_r  \\
y_i 
\end{bmatrix}$$

and essential dofs are eliminated from both $x_r$ and $x_i$.

Hope this helps.

Socratis

**Comment by najlkin** [comment]
Or you may use the new class `DarcyForm` designed exactly for such saddle-point problems in #4350 . However, I essentially do (in `FormLinearSystem()`) what you did, so maybe you have just some bug somewhere. Also check #4356 , which offers more methods for elimination of BCs in mixed forms and should get to `master` much sooner than the other PR :wink: .

**Comment by AIBSCT80** [comment]
@najlkin @psocratis 

many many thanks for your  help.  I looked into the code carefully.  Actually I have found that 

```
ufespace->GetBoundaryTrueDofs(u_ess_tdof_list);  
pfespace->GetBoundaryTrueDofs(p_ess_tdof_list);  
```

That is I assigned Dirichlet to dual variable `p` as well. That is a bug in the code, due to copy 
and paste.  

I comment out the this line, it seems to produce no zeros for the ` p` variable now.  

Thank you all for the suggests.
